addons/product/product.py

```diff
@@ -1069,26 +1069,26 @@ def name_search(self, cr, user, name='', args=None, operator='ilike', context=No
             positive_operators = ['=', 'ilike', '=ilike', 'like', '=like']
             ids = []
             if operator in positive_operators:
-                ids = self.search(cr, user, [('default_code','=',name)]+ args, limit=limit, context=context)
+                ids = self.search(cr, user, ['|', ('default_code', '=', name), ('attribute_value_ids.name', operator, name)] + args, limit=limit, context=context)
                 if not ids:
-                    ids = self.search(cr, user, [('barcode','=',name)]+ args, limit=limit, context=context)
+                    ids = self.search(cr, user, [('barcode', '=', name)] + args, limit=limit, context=context)
             if not ids and operator not in expression.NEGATIVE_TERM_OPERATORS:
                 # Do not merge the 2 next lines into one single search, SQL search performance would be abysmal
                 # on a database with thousands of matching products, due to the huge merge+unique needed for the
                 # OR operator (and given the fact that the 'name' lookup results come from the ir.translation table
                 # Performing a quick memory merge of ids in Python will give much better performance
-                ids = self.search(cr, user, args + [('default_code', operator, name)], limit=limit, context=context)
+                ids = self.search(cr, user, args + ['|', ('default_code', operator, name), ('attribute_value_ids.name', operator, name)], limit=limit, context=context)
```

Comment:
Don't modify this one either, it's still matching on default_code only.
The only search calls you can modify are the ones searching on `name`, i.e. lines 1084 and 1086. 
And when you modify it, you should search in the display_name of the variants, as going through a `to-many` relationship (`attribute_value_ids`) will be very bad in terms of performance.

As you can see in the comments, performance is critical in the `name_search` of products. You can test your patch in a database with 50k products and see if it is still fast.
Thanks!